// initialization/SensorTypes.js
const dbcalls = require('../utility_modules/database_calls')
const i2c_addresses = require('../../config/i2c_addreses.json')

module.exports.createThermometerOpts = function createThermometerOpts(currentState, DS18B20_Array, DS18B20_data, obj){
  // If DS18B20, go through array of addresses
  if(currentState.sensorModel == "DS18B20"){     
    let address = currentState.sensorAddress
    // If address is null, pull from bottom of array and update sensor
    if (address == null) {
      address = DS18B20_Array[0]
      dbcalls.updateSensorAddress(address, currentState.sensorID)                    
    }
    //Remove address from database
    const index = DS18B20_Array.indexOf(address);
    if (index > -1) {
      DS18B20_Array.splice(index, 1);
    }
    obj.pin = DS18B20_data.pin
    address = Number(address)
    obj.address = address
  } else {
    //Otherwise normal temperature sensor
  } 
  return obj
}

module.exports.createHygrometerOpts = function createHygrometerOpts(obj){
  return obj
}

module.exports.createBarometerOpts = function createBarometerOpts(obj){
  return obj
}


//Gets DS18B20 addresses from the board, assigning them to array for adding to the database. 
module.exports.DS18B20AddressFilter = function DS18B20AddressFilter(DS18B20_data, state, addr) {
  //If this address exists in database, do not include in list
  let include = true;
  for (i = 0; i < state.sensorState.getSensorState().length; i++) {
    if (state.sensorState.data[i].sensorModel == "DS18B20" && state.sensorState.data[i].sensorAddress == addr) {              
      include = false;
    }
  }
  if (include){
    DS18B20_data.array.push(addr)
  }
  DS18B20_data.count--
}


//Gets pin and count for DS18B20 sensors
module.exports.DS18B20Initializer = function DS18B20Initializer(state){
  // Get DS18B20 count and pin
  let count = 0;
  let pin = -1
  for (let i = 0; i < state.sensorState.getSensorState().length; i++) {      
    if (state.sensorState.data[i].sensorModel == "DS18B20") {
      count++
      pin = state.sensorState.data[i].sensorPin
    }
  }
  return { count, pin }
}

//Adds addresses to I2C devices
module.exports.I2CFilter = function I2CFilter(obj, currentState){
    if(currentState.sensorModel == "BME280"){
      obj.address = Number(i2c_addresses.BME280_ADDRESS);     
    }  
}
