// initialization/SystemInitializer.js
const five = require("johnny-five");
const mockFirmata = require("mock-firmata")
const dbcalls = require('../utility_modules/database_calls.js')
const events = require('events');
const pEventHandler = require('../event_modules/event_handlers/PeriodicEventHandler.js');
const sEventHandler = require('../event_modules/event_handlers/SensorEventHandler.js');
const tEventHandler = require('../event_modules/event_handlers/TimeEventHandler.js');
const sLogger = require('../event_modules/SensorLogger.js');
const SensorTypes = require('../initialization/SensorTypes.js')
const utils = require('../utility_modules/Utils.js');
if(process.env.NODE_ENV == 'test'){
  var config = require('../../config/test-config.json');
} else {  
  var config = require('../../config/config.json');
}

const LOGGING_TIMER = 60 * 1000 // 1 MINUTE
const EVENT_TIMER = 1 * 1000 // 1 SECOND
var board;
let DS18B20_data = {}
DS18B20_data.count = 0;
DS18B20_data.pin = -1;
DS18B20_data.array = [];
var relay_control;

module.exports.initialize = async function initialize(state, app) {
  const doneEmitter = new events.EventEmitter();
       
  doneEmitter.once('booted', async () => {
      console.log("Initializing sensors. . .");
      await initializeSensors(state, DS18B20_data)
      console.log("Initializing outputs. . ."); 
      await initializeOutputs(state, config)
      doneEmitter.emit('hardwareInitialized', state)
  })

  doneEmitter.once('hardwareInitialized', async (state) => {
    console.log("Initializing schedule. . .")
    await initializeSchedule(state)
    state.warnState = app.get('warnState')
    app.set('state', state)
    // If relay control exists, turn on relays
    if(relay_control){
      relay_control.high()
    }       
    console.log(". . .Done!");
    app.emit('started');
  })
  
  return new Promise(async (resolve) => {
    let DS18B20_result = SensorTypes.DS18B20Initializer(state) 
    DS18B20_data.count = DS18B20_result.count
    DS18B20_data.pin = DS18B20_result.pin  
    console.log("Initializing board. . .");  
    board = createBoardObject(config)
    state.board = board;
    resolve(board)
  }).then((board) => {
    //Find and push DS18B20 Address onto Array
    board.once("ready", function() {             
      if(config.board == 'test'){
        doneEmitter.emit('booted')
      } else {
        if (DS18B20_data.count != 0) {
          //Check board drivers for DS18B20's on the given pin
          five.Thermometer.Drivers.get(
            this, "DS18B20", {
            pin: DS18B20_data.pin
          }).on('initialized', function(addr){
            //For each DS18B20 found on initialization, do stuff with it
            SensorTypes.DS18B20AddressFilter(DS18B20_data, state, doneEmitter, addr)
            //If no more DS18B20, emit
            if (DS18B20_data.count == 0 ){            
              doneEmitter.emit('booted')
            }   
          })    
        }
        //If no DS18B20, emit      
        doneEmitter.emit('booted')
      }      
    })
  });
}

async function initializeOutputs(state, config) {
  return new Promise((resolve) => {
    //If relay toggle prevention is set in config, create pin with first output pin
    if(config.relay_toggle_prevention){
      let relay_toggle_pin = utils.getBoardMapping().START_OUTPUT_PIN
      relay_control = new five.Pin(relay_toggle_pin)
      //Handle board interrupts (turn off stuff)
      board.on('exit', () => {
        relay_control.low()           
      })
    }
    //Set up outputs and bind to state object.
    for(let i = 0; i < state.outputState.getOutputState().length; i++){
      let outputPin = state.outputState.data[i].outputPin
      var output = new five.Relay({
        pin: outputPin,
        type: "NC"
      })
      state.outputState.data[i].outputObject = output
      if(state.outputState.data[i].outputPWMPin != null){
        let PWMPin = state.outputState.data[i].outputPWMPin
        var PWM = new five.Led(PWMPin);
        state.outputState.data[i].outputPWMObject = PWM
        //Handle board interrupts (turn off stuff)
        board.on('exit', () => {
          output.open()
          PWM.brightness(0)
        })
      } else {
        //Handle board interrupts (turn off stuff)
        board.on('exit', () => {
          output.open()           
        })
      }
      try {
        output.open()
      } catch (e) {
        utils.errorPrintout("Arduino out of pins! Could not attach all outputs, please check your settings.")
      }
    }
    resolve();
  });
}
module.exports.initializeOutputs = initializeOutputs

async function initializeSensors(state, DS18B20_data) {
  return new Promise((resolve) => {
    //Set up sensors and bind to state object.
    let DS18B20_Array = DS18B20_data.array.map(e => e.toString())
    //Get max hardwareID
    var hardwareID = Math.max.apply(Math, state.sensorState.data.map(function(o) { return o.sensorHardwareID; }));
    //Loop through hardware IDs
    let counter = 0
    while (hardwareID > counter) {
      counter++
      let i = 0;
      //Loop through sensors to find ones with matching hardwareID
      for(i; i < state.sensorState.getSensorState().length; i++){  
        let currentState = state.sensorState.data[i]     
        let sensor;
        obj = {controller: currentState.sensorModel}
        SensorTypes.I2CFilter(counter, obj, currentState)
        // Else if the hardware ID is the current ID, get pin   
        if(counter == currentState.sensorHardwareID) {
          obj.pin = currentState.sensorPin
        // Otherwise, skip
        } else {
          continue
        }
        // Create sensor based on type
        switch (currentState.sensorType) {
          case 'Temperature':
            sensor = new five.Thermometer(SensorTypes.createThermometerOpts(currentState, DS18B20_Array, DS18B20_data, obj))
            break;
          case 'Humidity':
            sensor = new five.Hygrometer(SensorTypes.createHygrometerOpts(obj))
            break;
          case 'Pressure':
            sensor = new five.Barometer(SensorTypes.createBarometerOpts(obj))
            break;
          case 'CarbonDioxide':
            sensor = -1;
            break;
          default:
            sensor = -1;
          }
          currentState.sensorObject = sensor;          
        }
      }
    resolve()
  });
}
module.exports.initializeSensors = initializeSensors

//Logic for event checking - checks once a minute
//Check once on load, then every minute thereafter.
async function initializeSchedule(state) {
  //Take initial reading to update database
  //Run events when ready, then set Interval.
  state.events = await dbcalls.getEnabledEvents()
  await sLogger.addSensorReadings(state)
  await utils.regenerateChartData()
  await tEventHandler.TimeEventHandler(state);
  await sEventHandler.SensorEventHandler(state);
  await pEventHandler.PeriodicEventHandler(state);
  setInterval(async function() {
    await sLogger.addSensorReadings(state)
    await utils.regenerateChartData()
  }, LOGGING_TIMER);
  setInterval(async function() {
    await utils.scheduleMinder(state)
    await tEventHandler.TimeEventHandler(state);
    await sEventHandler.SensorEventHandler(state);
    await pEventHandler.PeriodicEventHandler(state);
  }, EVENT_TIMER);
}
//Creates the board object for initialization
function createBoardObject(config){
  let board_object = {
    repl: false,
    // debug: false
  }
  if(config.board == 'test'){
    board_object.debug = false
    board_object.io = new mockFirmata.Firmata()
    board_object.io.emit("connected")
    board_object.io.emit("ready")
  }
  let board = new five.Board(board_object);
  return board
}
module.exports.createBoardObject = createBoardObject
