// utility_modules/config_check.js
const fs = require('fs');
const { data } = require('jquery');
const readline = require('readline')
const {simpleLogPrintout, simpleErrorPrintout} = require('./printouts')

const getInput = function(query) {
  const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
  });

  return new Promise(resolve => rl.question(query, ans => {
      rl.close();
      resolve(ans);
  }))
}
module.exports.getInput = getInput

// returns config.json from the config directory; if not, it copies it from default_config and returns it.
const initializeConfig = async function() {
  var config;
  return new Promise(async (resolve, reject) => {
    if(fs.existsSync('./config/config.json')){
      config = JSON.parse(fs.readFileSync('./config/config.json').toString())
      resolve(config)
    } else {
      simpleLogPrintout("[SETUP] Could not locate config.json, copying from default_config.json.");
      try {
        fs.copyFileSync('./config/default_config.json', './config/config.json')
        config = JSON.parse(fs.readFileSync('./config/config.json').toString())
        resolve(config)
      } catch(err) {
        simpleErrorPrintout("ERROR: Could not copy config.json from default_config.json.")
        reject(err)
      }
    }
  });    
}
module.exports.initializeConfig = initializeConfig

// Gets a list of the boards available, returns them as both an object and a string.
const getBoardData = function() {
  let board_list_JSON;
  let board_list = ''
  try {
    board_list_JSON = JSON.parse(fs.readFileSync('./config/board_mappings.json').toString())
  } catch (e) {
    simpleErrorPrintout('ERROR: Could not load board mappings.')
    process.exit(-1)
  }
  let i = 1
  for (let e in board_list_JSON){
    board_list = board_list.concat(e)
    if (i != Object.keys(board_list_JSON).length){
      board_list = board_list.concat(", ")
      i++;
    }
  }
  return [board_list_JSON, board_list]
}
module.exports.getBoardData = getBoardData

//Updates config with the passed version.
const configUpdate = function(config, update) {
  if (update == true){
    try {      
      fs.writeFileSync('./config/config.json', JSON.stringify(config))
      simpleLogPrintout("\n[SETUP] Config has been updated!\n")
      return true
    } catch (e) {
      simpleErrorPrintout('ERROR: Could not update config.')
      process.exit(-1)
    }
  }
  return false
}
module.exports.configUpdate = configUpdate

// Checks config to see if there's a board listed - updates config with input.
const boardCheck = async function(config, update){
  if(config.board == 'null'){
    let [ board_list_JSON, board_list ] = exports.getBoardData()
    simpleLogPrintout("[SETUP] No board specified in config. Please enter your arduino model:")
    let board;
    while (!(board in board_list_JSON)){
      simpleLogPrintout("Available: " + board_list);        
      board = await exports.getInput('> ')
    }
    config.board = board
    update = true 
  } 
  return [config, update]
}
module.exports.boardCheck = boardCheck

const databaseCheck = async function(config, update) {
  if (config.database.password == 'null'){
    simpleLogPrintout("\n[SETUP] No database password specified in config. Please enter the database password for growth_admin:")
    config.database.password = await exports.getInput("> ")  
    update = true  
  }
  return [config, update]
}
module.exports.databaseCheck = databaseCheck

const jwtCheck = async function(config, update) {
  if (config.jwt_secret == 'null'){
    simpleLogPrintout("\n[SETUP] jwt_secret has not been changed in config. Please enter a string to use for authentication security:")
    config.jwt_secret = await exports.getInput("> ")
    update = true
  }
  return [config, update]
}
module.exports.jwtCheck = jwtCheck

const nodemailerCheck = async function(config, update, warnState) {
  if ((config.nodemailer.service == 'null' || config.nodemailer.auth.user == 'null' || config.nodemailer.auth.pass == 'null') && config.nodemailer_setup_warn == 'true'){
    simpleLogPrintout("\n[SETUP] nodemailer has not been setup in config, 'Warn' will not be sent.")
    let input = await exports.getInput("Would you like to set this up now? (y/N) ")
    if (input == 'y' || input == 'Y'){
      config.nodemailer.service = await exports.getInput("Please input an emailing service (gmail): \n> ")
      config.nodemailer.auth.user = await exports.getInput("Please input the account's email address: \n>  ")
      config.nodemailer.auth.pass = await exports.getInput("Please input the account's password: \n>  ")
      config.nodemailer_setup_warn = false
      update = true
    } else {
      input = await exports.getInput("Would you like to disable 'Warn' setup in the future? (Y/n) ")
      if (input == 'n' || input == 'N'){
        config.nodemailer_setup_warn = false
        update = true
      }
      warnState = false;
    }
  }
  return [config, update, warnState]
}
module.exports.nodemailerCheck = nodemailerCheck

const relayTogglePreventionCheck = async function(config, update) {
  if (config.relay_toggle_prevention == 'null'){
    simpleLogPrintout("\n[SETUP] Relay toggle prevention has not been changed in config.")
    input = await exports.getInput("Would you like to enable this? (y/N)")
    if (input == 'y' || input == 'Y'){
      config.relay_toggle_prevention = true
    } else {
      config.relay_toggle_prevention = false
    }
    update = true
  }
  return [config, update]
}
module.exports.relayTogglePreventionCheck = relayTogglePreventionCheck

const configChecker = async function(config, warnState){
  let update = false;
  return new Promise(async (resolve, reject) => {
    try {
      [config, update] = await exports.boardCheck(config, update);
      [config, update] = await exports.databaseCheck(config, update);
      [config, update] = await exports.jwtCheck(config, update);    
      [config, update, warnState] = await exports.nodemailerCheck(config, update, warnState);
      [config, update] = await exports.relayTogglePreventionCheck(config, update);          
      resolve()
    } catch (e) {
      reject(e)
    }
  }).then(() => {
    exports.configUpdate(config, update)
    return true   
  }).catch((e) => {
    simpleErrorPrintout("ERROR: Could not update config.")
    if (e instanceof TypeError) {
      simpleErrorPrintout('Config may be corrupted. Please fix or delete, then try again.')
    }
    process.exit(-1)
  })
}

module.exports.configChecker = configChecker
