// utility_modules/Mappings.js
const fs = require('fs')
var config_helper = require('./config_helper');
const dbcalls = require('./database_calls.js')
const {errorPrintout} = require('./printouts')

const BOARD_MAPPINGS_LOCATION = './config/board_mappings.json'
var board = null;
/**
 * Singleton; Gets a JSON pinout dictionary for a Board. Creates it if it does
 * not exist.
 * @returns JSON pinout for the board listed in config.
 */
module.exports.getBoardMapping = function getBoardMapping(){
  var config = config_helper.getConfig()
  if(board == null) {
    let board_mappings = JSON.parse(fs.readFileSync(BOARD_MAPPINGS_LOCATION).toString())
    board = board_mappings[config.board]  
  }
  return board
}

//Maps pins for each output.
/**
 * Gets a mapping of outputs to output and PWM pins referencing the database.
 * @returns {[object]} A mapping of outputs to output and PWM pins.
 */
module.exports.getOutputMappings = async function getOutputMapings() {
  return exports.mapOutputPins(await dbcalls.getEnabledOutputs())
}

//Maps pins for each sensor
/**
 * Gets a mapping of sensors to the appropriate pins referencing the database.
 * @returns {[object]} A mapping of sensors to various data pins.
 */
module.exports.getSensorMappings = async function getSensorMappings() {
  return exports.mapSensorPins(await dbcalls.getEnabledSensors())
}

/**
 * Creates a mapping of outputs to output and PWM pins. Adds pins to the passed
 * object.
 * @param {[object]} outputs An array of outputs and their properties.
 * @returns {[object]} A mapping of outputs to output and PWM pins.
 */
function mapOutputPins(outputs) {
  var config = config_helper.getConfig()
  var pinout = exports.getBoardMapping()
  var outputPin = pinout.START_OUTPUT_PIN;
  var pwmPin = pinout.START_PWM_PIN
  //If relay toggle prevention is set in config, skip first output
  if(config.relay_toggle_prevention){
    outputPin++;
  }
  //Iterate through outputs
  for(i = 0; i < outputs.length; i++){
    if (outputPin > pinout.END_OUTPUT_PIN) {
      errorPrintout("Mappings.js: mapOutputPins() failed, out of digital pins!");
      process.exit(-1)
      return
    }
    //Otherwise set OUTPUTPIN property to outputPin and increment to the next one
    outputs[i].OUTPUT_PIN = outputPin;
    outputPin++;
    //If we've run out of PWM pins, error.
    if (pwmPin > pinout.END_PWM_PIN) {
      errorPrintout("Mappings.js: mapOutputPins() failed, out of PWM pins!");
      process.exit(-1)
      return
    }
    //Otherwise apply PWMPIN property if necessary.
    if(outputs[i].outputPWM == 1) {
        outputs[i].PWM_PIN = pwmPin;
        pwmPin++;
    }
  };
  return outputs;
}
module.exports.mapOutputPins = mapOutputPins

/**
 * Creates a mapping of sensors to the appropriate pins. Adds pins to the passed
 * object.
 * @param {[object]} outputs An array of sensors and their properties.
 * @returns {[object]} A mapping of sensors to various data pins.
 */
function mapSensorPins(sensors) {
  var pinout = exports.getBoardMapping()
  let sensorPin = pinout.START_SENSOR_PIN;
  let analogPin = pinout.START_ANALOG_PIN;
  let onewirePin = -1;
  //Iterate through sensors
  let hardwareObject = {}
  for(i = 0; i < sensors.length; i++) {
    //Check to see if hardwareID exists in object
    if(Object.keys(hardwareObject).indexOf(String(sensors[i].sensorHardwareID)) != -1){
      //if yes, set to stored hardwareID and end.
      sensors[i].SENSOR_PIN = hardwareObject[sensors[i].sensorHardwareID]
      continue;
    }
  
    //If we've run out of sensor pins, error.
    if (sensorPin > pinout.END_SENSOR_PIN || analogPin > pinout.END_ANALOG_PIN) {
      errorPrintout("Mappings.js: mapSensorPins() failed, out of pins!");
      process.exit(-1)
      break;
    }
    //Otherwise, case switch to apply pins to appropriate sensors
    switch (sensors[i].sensorProtocol) {
      case "I2C":
        continue;
      case "ANALOG":
        hardwareObject[sensors[i].sensorHardwareID] = analogPin;
        sensors[i].SENSOR_PIN = analogPin;
        analogPin++;
        continue;
      case "ONEWIRE":
        if (onewirePin == -1) {
          onewirePin = sensorPin
          sensorPin++;
        }
        hardwareObject[sensors[i].sensorHardwareID] = onewirePin;
        sensors[i].SENSOR_PIN = onewirePin;
        continue;
      default: 
        hardwareObject[sensors[i].sensorHardwareID] = sensorPin;
        sensors[i].SENSOR_PIN = sensorPin;
        sensorPin++;
    } 
  };
  return sensors;
}
module.exports.mapSensorPins = mapSensorPins
